internal/wrphandlers/qos/priority_queue.go

```
@@ -0,0 +1,101 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"container/heap"
+
+	"github.com/xmidt-org/wrp-go/v3"
+)
+
+// priorityQueue implements heap.Interface and holds wrp Message, using wrp.QOSValue as its priority.
+// https://xmidt.io/docs/wrp/basics/#qos-description-qos
+type priorityQueue struct {
+	// queue for wrp messages, ingested by serviceQOS
+	queue []wrp.Message
+	// maxQueueSize is the allowable max size of the queue based on the sum of all queued wrp message's payloads
+	maxQueueSize int
+	// MaxMessageBytes is the largest allowable wrp message payload.
+	maxMessageBytes int
+}
+
+// Dequeue returns the next highest priority message.
+func (pq *priorityQueue) Dequeue() (wrp.Message, bool) {
+	// Required, otherwise heap.Pop will panic during an internal Swap call.
+	if pq.Len() == 0 {
+		return wrp.Message{}, false
+	}
+
+	msg, ok := heap.Pop(pq).(wrp.Message)
+
+	return msg, ok
+}
+
+// Enqueue queues the given message.
+func (pq *priorityQueue) Enqueue(msg wrp.Message) {
+	// Check whether msg violates maxMessageBytes.
+	if len(msg.Payload) > pq.maxMessageBytes {
+		return
```

Dropping the incoming message because it's too large should return an error.  This is different than the message (or any message) being dropped due to being lowest priority.